### ISSUE TITLE
Loss diverges at the 3rd training step between GroupedGEMM and Sequential GEMMs

### ISSUE THREAD

**Issue by fanshiqing** [tgale96/grouped_gemm#1]
Hi @tgale96 , yesterday I successfully replacing the sequential gemm in SwitchMLP in Megatron-LM with your developed grouped gemm. But find that the loss diverges at the 3rd training step between GroupedGEMM and Sequential GEMMs, and want to seek help from you if you have any debug idea?

Takeaway:
- Model: GPT3-2.3B/8E, training, w/ swiglu, TP4DP1PP1 w/ SequenceParallel, A100x4, each expert is distributed across 4 TP ranks;
- The fwd activation outputs of GroupedGEMM and Sequential GEMMs of the first two training step are exactly equal (ensured by  `assert torch.allclose(output_total, output_total_gmm, rtol=1e-5, equal_nan=True)`, while since the third step, the loss diverges.
- Loss curve: 
![image](https://github.com/tgale96/grouped_gemm/assets/3927162/c000914b-434e-4e8f-b8e3-e8b869527cab)

**Comment by fanshiqing** [comment]
I noted that in Megablocks it claims that "Weight parallelism not yet supported with GroupedMLP" ([src](https://github.com/stanford-futuredata/megablocks/blob/main/megablocks/layers/mlp.py#L516-L518)), and in my case each expert does sharded across TP rank, I am wondering if this will introduce any risk which finally introduce loss diverge?

**Comment by fanshiqing** [comment]
Updates: it seems this problem is **not related to the TP sharding**, because I have already disable the TP sharding while the accuracy diverge still happens at the third iteration.

**Comment by fanshiqing** [comment]
For ease of observation, I calculate the percentage of elements that not close (error > 1e-5) from 8192x2048 activation tensor from SequentialGEMM and GroupedGEMM respectively, and I got the figure as below:

![image](https://github.com/tgale96/grouped_gemm/assets/3927162/c6e3ccd9-7b67-4865-91f0-8422b35e2f3a)

- The first two iterations generate exact the same output
- Then the PCT of elements gradually grows larger and larger (grows to >92% after 100 iters)

Is this expected behavior of CUTLASS GroupedGEMM?

and I find that the [UT](https://github.com/tgale96/grouped_gemm/blob/main/grouped_gemm/ops_test.py#L10-L17) from our GroupedGEMM also assumes the <2% of different elements is tolerable.

**Comment by fanshiqing** [comment]
Updates: sorry for the misleading, there are something wrong with my Debug's code, now I verified that the GroupedGEMM works as expected compared to the Sequential GEMMs.

**Comment by tgale96** [comment]
Glad you were able to figure it out!
